ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala

```diff
@@ -61,70 +63,71 @@ private[apiserver] final class LedgerTimeAwareCommandExecutor(
+                )
+                success(advanceOutputTime(cer, maxUsedTime))
+
+              case Success(MaximumLedgerTime.Max(maxUsedTime)) =>
+                if (retriesLeft > 0) {
                   logger.debug(
                     s"Restarting the computation with new ledger effective time $maxUsedTime"
                   )
-                  val advancedCommands = advanceInputTime(commands, maxUsedTime)
-                  loop(advancedCommands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
+                  retry(advanceInputTime(commands, maxUsedTime))
                 } else {
-                  Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
+                  failed
                 }
-              }
-              .recoverWith {
-                case MissingContracts(contracts) =>
-                  if (retriesLeft > 0) {
-                    metrics.daml.execution.retry.mark()
-                    logger.info(
-                      s"Some input contracts could not be found. Restarting the computation. Missing contracts: ${contracts
-                        .mkString("[", ", ", "]")}"
-                    )
-                    loop(commands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
-                  } else {
-                    logger.info(
-                      s"Lookup of maximum ledger time failed after ${maxRetries - retriesLeft}. Used contracts: ${usedContractIds
-                        .mkString("[", ", ", "]")}."
-                    )
-                    Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
-                  }
-
-                // An error while looking up the maximum ledger time for the used contracts. The nature of this error is not known.
-                // Not retrying automatically. All other automatically retryable cases are covered by the logic above.
-                case error =>
+
+              case Success(Archived(contracts)) =>
+                if (retriesLeft > 0) {
+                  logger.info(
+                    s"Some input contracts are archived: ${contracts.mkString("[", ", ", "]")} Restarting the computation."
+                  )
+                  retry(commands)
+                } else {
                   logger.info(
                     s"Lookup of maximum ledger time failed after ${maxRetries - retriesLeft}. Used contracts: ${usedContractIds
-                      .mkString("[", ", ", "]")}. Details: $error"
+                      .mkString("[", ", ", "]")}."
                   )
-                  Future.successful(Left(ErrorCause.LedgerTime(maxRetries - retriesLeft)))
-              }
+                  failed
+                }
+
+              // An error while looking up the maximum ledger time for the used contracts. The nature of this error is not known.
+              // Not retrying automatically. All other automatically retry-able cases are covered by the logic above.
+              case Failure(error) =>
```

Comment:
This is an internal error (such as a transient database error), right? Should we fail the whole `execute` call here with a self-service error message?

Reply:
this is exactly what is going to happen at https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala#L174